### PR TITLE
Preserve data

### PR DESCRIPTION
**Pull request by Apatakiti**
## Project requirements

**Check the documentation for Ruby's JSON library**.

- On exit from your app save all the data (books, people, and rentals) into JSON files. One per type of data you are storing and named accordingly (e.g. rentals.json)
- On startup of your app load all data from files. Remember to handle the case when one or more of the files are missing.

### CODE REVIEW

File: main.rb
Comment (Efiamotu-1) [inline main.rb:1]: - [x] When i tried running your code to test that it is in line with the `project milestone requirements`, i get the below error as shown in the image below, please kindly fix this error and resubmit your code :+1: :sparkles: 
nerr

File: README.md
Comment (Efiamotu-1) [inline README.md:1]: - [x] I kindly suggest you add the information about the other coding partner to the authors section of your readme seeing that it is a pari programming project :+1:

**Review by Efiamotu-1 — CHANGES_REQUESTED** [review]
Hi @Team, 

 

Good job so far! 👏. Your Code is Awesome! 💪

There are some **few** issues 🐛 that you still need to work 🔧 on though to go to the next project ⏭️ but you are almost there! 🏃
## Don't Worry 🙏

It's ok to have bugs in your code 🐛. Please, see this code review as a **chance** for you to **learn** practical coding skills and learn **new tips** that can improve your code quality 💯 and help you **get a better ** 💼.

_I am going to guide you on how to fix them down below. 🤓_

## Highlights :dart: 
- ✔️ No linter Errors
- ✔️ Descriptive README
- ✔️ Descriptive PR title and description
- The code is top-notch! 💯 and has a huge potential.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, don't Merge This Pull Request yet. Only Merge when it is Approved ✅.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: main.rb
Comment (Apatakiti) [reply main.rb:1]: fixed and corrected

File: README.md
Comment (Apatakiti) [reply README.md:1]: Done

File: data_preserver.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      if person.instance_of?(Teacher)
+        data.push({ id: person.id, age: person.age, name: person.name,
+                    specialization: person.specialization, data_type: person.class })
+      else
+        data.push({ id: person.id, age: person.age, name: person.name,
+                    parent_permission: person.parent_permission, data_type: person.class })
+      end
+    end
+    File.write('people.json', JSON.generate(data))
+  end
```
Comment (1GORDON) [inline data_preserver.rb]: Great job team🤝
- As stated in the project requirements, on exit from the student's app check the data is being saved (all the data: books, people, and rentals) into JSON files. Which is not the case for your application. I am instead getting this error. Kindly fix this issue and endeavor to make sure your app meets the intended requirements.
error

**Review by 1GORDON — CHANGES_REQUESTED** [review]
##
## Hi @Ademola-coding, and @fmanimashaun  ✋

Good job so far!

## Highlights!
- Updated the readme file✔️

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: data_preserver.rb (hunk 1 above)
Comment (Apatakiti) [reply data_preserver.rb]: **corrected and fixed**

Screenshot (91)

**Review by shella12 — COMMENTED** [review]
Hi, 

#  Status: Invalidated :red_circle: 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a PR from previous mile stone kindly create a new PR for this milestone and make sure that your partner has significant contribution in the project (Not updating README, removing lines etc) as this is a pair-programming activity.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: app.rb [hunk 2]
```diff
@@ -1,59 +1,197 @@
...
-require './book'
-require './rental'
+require 'json'
+require './database'
+require './read_database'
+require_relative 'student'
+require_relative 'teacher'
+require_relative 'classroom'
+require_relative 'book'
+require_relative 'rental'
```
Comment (Munsa1) [inline app.rb]: -  [ ]  Good job on trying to fix the required changes, however you have introduced new changes that have changed how the app functions, kindly fix this

Video Demo
[cinnamon-2023-11-18T085113+0200.webm](https://github.com/Ademola-coding/School-Library/assets/81932478/fc3f6a51-0679-4410-bdab-b44824f22a82)


#### Notice how the app terminates after listing people

File: book.rb [hunk 3]
```diff
@@ -1,17 +1,14 @@
...
   end
 
-  def add_rental(date, person)
-    rental = Rental.new(date, person, self)
-    @rentals << rental unless rentals.includes?(rental)
+  def add_rental(rental)
+    @rentals.push(rental)
+    rental.book = self
   end
 end
```
Comment (Munsa1) [inline book.rb:14]: - [ ] Also after creating the book, when the user lists all books, the added book does not show and the app terminates, kindly fix this as it has been introduced with the new commits

Screenshot
image


- The program should not exit after creating a book and a person, kindly fix this as it has been introduced with the new commits
Before new commits
image

File: person.rb [hunk 4]
```diff
@@ -1,28 +1,23 @@
...
     @age = age
     @parent_permission = parent_permission
     @rentals = []
   end
 
-  private
-
-  def of_age?
-    true if @age >= 18
-    false
```
Comment (Munsa1) [inline person.rb:21]: - [ ] After creating a person, checking the list of people does not show said person, this is a bug that has been introduced with the new commits, kindly fix this.

Before commits
image

After
image

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @Team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Passing Linters :heavy_check_mark: 
Correct GitHub Flow :heavy_check_mark: 
Descriptive ReadMe File :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb (hunk 3 above)
Comment (Apatakiti) [reply book.rb:14]: Bugs identified and are all fixed

File: person.rb (hunk 4 above)
Comment (Apatakiti) [reply person.rb:21]: Bugs identified and are all fixed

File: book.rb (hunk 3 above)
Comment (Apatakiti) [reply book.rb:14]: **Adding Book and person**

Screenshot (96)

Screenshot (97)

**Book Added in JSON file**

Screenshot (99)

File: person.rb (hunk 4 above)
Comment (Apatakiti) [reply person.rb:21]: **ADDING PERSON**

Screenshot (96)

Screenshot (97)

**PERSON ADDED To JSON FILE**
Screenshot (98)

**Review by prolajumokeoni — APPROVED** [review]
Hi @Ademola-coding and @fmanimashaun 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
 - Data is preserved ✅ 
 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**